frontend/src/metabase/visualizations/visualizations/SmartScalar/utils.ts

```diff
@@ -63,3 +79,244 @@ export const getValueHeight = (height: number): number => {
+  | SelectedComparisonPeriodsAgo;
+
+export function getDefaultComparison(
+  series: RawSeries,
+  settings: VisualizationSettings,
+) {
+  const [
+    {
+      data: { insights },
+    },
+  ] = series;
+
+  const dateUnit = insights?.find(
+    insight => insight.col === settings["scalar.field"],
+  )?.unit as RelativeDatetimeUnit | undefined;
+
+  if (!dateUnit) {
+    return createComparisonMenuOption({
+      type: COMPARISON_TYPES.PREVIOUS_VALUE,
+    });
+  }
+
+  return createComparisonMenuOption({
+    type: COMPARISON_TYPES.PREVIOUS_PERIOD,
+    dateUnit,
+  });
+}
+
+export function getComparisonOptions(
+  series: RawSeries,
+  settings: VisualizationSettings,
+) {
+  const [
+    {
+      data: { cols, insights, rows },
+    },
+  ] = series;
+
+  const dateUnit = insights?.find(
+    insight => insight.col === settings["scalar.field"],
+  )?.unit as RelativeDatetimeUnit | undefined;
+
+  if (!dateUnit) {
+    return [
+      createComparisonMenuOption({ type: COMPARISON_TYPES.PREVIOUS_VALUE }),
+    ];
+  }
+
+  const options: ComparisonMenuOption[] = [
+    createComparisonMenuOption({
+      type: COMPARISON_TYPES.PREVIOUS_PERIOD,
+      dateUnit,
+    }),
+  ];
+
+  const maxPeriodsAgo = getMaxPeriodsAgo({ cols, rows, dateUnit });
+
+  // only add this option is # number of selectable periods ago is >= 2
+  // since we already have an option for 1 period ago -> PREVIOUS_PERIOD
+  if (maxPeriodsAgo && maxPeriodsAgo >= 2) {
+    options.push(
+      createComparisonMenuOption({
+        type: COMPARISON_TYPES.PERIODS_AGO,
+        dateUnit,
+        maxValue: maxPeriodsAgo,
+      }),
+    );
+  }
+
+  options.push(
+    createComparisonMenuOption({ type: COMPARISON_TYPES.PREVIOUS_VALUE }),
+  );
+
+  return options;
+}
+
+export function isComparisonValid(
+  series: RawSeries,
+  settings: VisualizationSettings,
+) {
+  const [
+    {
+      data: { insights },
+    },
+  ] = series;
+
+  if (
+    settings["scalar.comparisons"]?.type === COMPARISON_TYPES.PREVIOUS_VALUE
+  ) {
+    return true;
+  }
+
+  const dateUnit = insights?.find(
+    insight => insight.col === settings["scalar.field"],
+  )?.unit;
+
+  if (!dateUnit) {
+    return false;
+  }
+
+  return true;
+}
+
+type getMaxPeriodsAgoParameters = {
+  cols: DatasetColumn[];
+  rows: RowValues[];
+  dateUnit: RelativeDatetimeUnit;
+};
+
+function getMaxPeriodsAgo({
+  cols,
+  rows,
+  dateUnit,
+}: getMaxPeriodsAgoParameters) {
+  const dimensionIndex = cols.findIndex(col => isDate(col));
+
+  if (dimensionIndex === -1) {
+    return null;
+  }
+
+  const latestNonEmptyRow = rows.findLast(row => !isEmpty(row[dimensionIndex]));
+  const earliestNonEmptyRow = rows.find(row => !isEmpty(row[dimensionIndex]));
+
+  if (latestNonEmptyRow === undefined || earliestNonEmptyRow === undefined) {
+    return null;
+  }
+
+  const latestNonEmptyDate = latestNonEmptyRow[dimensionIndex] as string;
+  const earliestNonEmptyDate = earliestNonEmptyRow[dimensionIndex] as string;
```

Comment:
Should we handle other `RowValue` types?

Reply:
Because we already checked that the column is a date column and checked that the rows we have are non-empty, we [know that the values are going to be strings](https://github.com/metabase/metabase/blob/trend-chart-improvements/milestone-2-support-arbitrary-date-comparison/frontend/src/metabase-lib/types/utils/isa.js#L105).